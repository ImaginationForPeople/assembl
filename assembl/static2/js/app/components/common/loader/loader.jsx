// @flow
import * as React from 'react';
import { I18n } from 'react-redux-i18n';
// Component imports
import LoadingIcon from '../icons/loadingIcon/loadingIcon';
import ErrorIcon from '../icons/errorIcon/errorIcon';
import NoDataIcon from '../icons/noDataIcon/noDataIcon';

type LoaderType = {
  loaderComponent: React.Node,
  description: string
};

export const LOADER_TYPE = {
  LOADING: 'LOADING',
  ERROR: 'ERROR',
  NO_DATA: 'NO_DATA'
};

export type Props = {
  /** Loader type of type LOADER_TYPE (LOADING, ERROR, NO_DATA) */
  type: $Keys<typeof LOADER_TYPE>
};

const Loader = ({ type }: Props) => {
  // Loaders Content
  const LoaderTypeLoading: LoaderType = {
    description: I18n.t('common.loader.loading'),
    loaderComponent: <LoadingIcon />
  };
  const LoaderTypeError: LoaderType = {
    description: I18n.t('common.loader.error'),
    loaderComponent: <ErrorIcon />
  };
  const LoaderTypeNoData: LoaderType = {
    description: I18n.t('common.loader.no-data'),
    loaderComponent: <NoDataIcon />
  };
  const LOADER_TYPE_CONTENT = {
    LOADING: LoaderTypeLoading,
    ERROR: LoaderTypeError,
    NO_DATA: LoaderTypeNoData
  };

  let loaderContent;
  switch (type) {
  case LOADER_TYPE.LOADING:
    loaderContent = LOADER_TYPE_CONTENT.LOADING;
    break;
  case LOADER_TYPE.NO_DATA:
    loaderContent = LOADER_TYPE_CONTENT.NO_DATA;
    break;
  default:
    loaderContent = LOADER_TYPE_CONTENT.ERROR;
  }
  const { description, loaderComponent } = loaderContent;

  return (
    <div className="custom-loader center">
      {loaderComponent}
      <p className="description center">{description}</p>
    </div>
  );
};

export default Loader;