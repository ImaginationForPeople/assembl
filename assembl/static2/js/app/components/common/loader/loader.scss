// Variables
$animation-speed: 1.4s;

@keyframes rotator {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(270deg);
  }
}

@keyframes dash {
  0% {
    stroke-dashoffset: 190;
  }

  50% {
    stroke-dashoffset: 47.5;
    transform: rotate(135deg);
  }

  100% {
    stroke-dashoffset: 190;
    transform: rotate(450deg);
  }
}

.custom-loader {
  .spinner {
    animation: rotator infinite $animation-speed linear;
    margin-bottom: 2rem;
    width: 60px;
  }

  .path {
    animation: dash infinite $animation-speed ease-in-out;
    stroke: $purple;
    stroke-dasharray: 190;
    transform-origin: center;
  }
}
