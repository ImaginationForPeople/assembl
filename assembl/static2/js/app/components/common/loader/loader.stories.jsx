// @flow
import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, select } from '@storybook/addon-knobs';
/* eslint-enable */

import Loader, { LOADER_TYPE } from '../../common/loader/loader';
import type { Props as LoaderProps } from '../../common/loader/loader';

export const defaultLoaderProps: LoaderProps = {
  type: LOADER_TYPE.LOADING
};

const loaderErrorProps: LoaderProps = {
  ...defaultLoaderProps,
  type: LOADER_TYPE.ERROR
};

const playground = {
  type: [LOADER_TYPE.LOADING, LOADER_TYPE.ERROR, LOADER_TYPE.NO_DATA]
};

storiesOf('Loader', module)
  .addDecorator(withKnobs)
  .add('default', withInfo()(() => <Loader {...defaultLoaderProps} />))
  .add('error', withInfo()(() => <Loader {...loaderErrorProps} />))
  .add('playground', withInfo()(() => <Loader type={select('type', playground.type, playground.type[0])} />));