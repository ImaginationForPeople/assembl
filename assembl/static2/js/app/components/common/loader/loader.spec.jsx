// @flow
import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.3';
/* eslint-enable */

import Loader, { LOADER_TYPE_CONTENT, LOADER_TYPE } from './loader';
import type { Props as LoaderProps } from './loader';
import { defaultLoaderProps } from './loader.stories';
import LoadingIcon from './icons/loadingIcon/loadingIcon';
import ErrorIcon from './icons/errorIcon/errorIcon';

configure({ adapter: new Adapter() });

describe('<Loader /> - with shallow', () => {
  let wrapper;
  let loaderProps: LoaderProps;

  describe('when is loading', () => {
    beforeEach(() => {
      loaderProps = { ...defaultLoaderProps };
      wrapper = shallow(<Loader {...loaderProps} />);
    });

    it('should render a loading icon', () => {
      expect(wrapper.find(LoadingIcon)).toHaveLength(1);
    });

    it('should render a loading description', () => {
      const { description } = LOADER_TYPE_CONTENT.LOADING;
      expect(wrapper.find('p').text()).toEqual(description);
    });
  });

  describe('when is error', () => {
    beforeEach(() => {
      loaderProps = {
        ...defaultLoaderProps,
        type: LOADER_TYPE.ERROR
      };
      wrapper = shallow(<Loader {...loaderProps} />);
    });

    it('should render an error icon', () => {
      expect(wrapper.find(ErrorIcon)).toHaveLength(1);
    });

    it('should render a loading description', () => {
      const { description } = LOADER_TYPE_CONTENT.ERROR;
      expect(wrapper.find('p').text()).toEqual(description);
    });
  });
});