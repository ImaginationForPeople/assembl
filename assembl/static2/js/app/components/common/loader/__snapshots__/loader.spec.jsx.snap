// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Loader default 1`] = `
<div>
  <div
    style={
      Object {
        "position": "relative",
        "zIndex": 0,
      }
    }
  >
    <div
      className="custom-loader center"
    >
      <svg
        className="icon spinner"
        viewBox="0 0 66 66"
      >
        <circle
          className="path"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          strokeLinecap="round"
          strokeWidth="6"
        />
      </svg>
      <p
        className="description center"
      >
        Chargement en cours...
      </p>
    </div>
  </div>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "#28c",
        "border": "none",
        "borderRadius": "0 0 0 5px",
        "color": "#fff",
        "cursor": "pointer",
        "display": "block",
        "fontFamily": "sans-serif",
        "fontSize": "12px",
        "padding": "5px 15px",
        "position": "fixed",
        "right": 0,
        "top": 0,
      }
    }
    type="button"
  >
    Show Info
  </button>
  <div
    style={
      Object {
        "background": "white",
        "bottom": 0,
        "display": "none",
        "left": 0,
        "overflow": "auto",
        "padding": "0 40px",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999,
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#28c",
          "border": "none",
          "borderRadius": "0 0 0 5px",
          "color": "#fff",
          "cursor": "pointer",
          "display": "block",
          "fontFamily": "sans-serif",
          "fontSize": "12px",
          "padding": "5px 15px",
          "position": "fixed",
          "right": 0,
          "top": 0,
        }
      }
      type="button"
    >
      ×
    </button>
    <div>
      <div
        style={
          Object {
            "WebkitFontSmoothing": "antialiased",
            "backgroundColor": "#fff",
            "border": "1px solid #eee",
            "borderRadius": "2px",
            "color": "#444",
            "fontFamily": "-apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", \\"Arial\\", sans-serif",
            "fontSize": "15px",
            "fontWeight": 300,
            "lineHeight": 1.45,
            "marginBottom": "20px",
            "marginTop": "20px",
            "padding": "20px 40px 40px",
          }
        }
      >
        <div
          style={
            Object {
              "borderBottom": "1px solid #eee",
              "marginBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <h1
            style={
              Object {
                "fontSize": "35px",
                "margin": 0,
                "padding": 0,
              }
            }
          >
            Loader
          </h1>
          <h2
            style={
              Object {
                "fontSize": "22px",
                "fontWeight": 400,
                "margin": "0 0 10px 0",
                "padding": 0,
              }
            }
          >
            default
          </h2>
        </div>
        
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Story Source
          </h1>
          <pre
            className="css-r8d96o eyvlbql0"
          >
            <div>
              <div
                style={
                  Object {
                    "paddingLeft": 18,
                    "paddingRight": 3,
                  }
                }
              >
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  &lt;
                  Loader
                </span>
                <span>
                  <span>
                     
                    <span
                      style={Object {}}
                    >
                      type
                    </span>
                    <span>
                      =
                      <span
                        style={Object {}}
                      >
                        "
                        <span
                          style={
                            Object {
                              "color": "#22a",
                              "wordBreak": "break-word",
                            }
                          }
                        >
                          LOADING
                        </span>
                        "
                      </span>
                    </span>
                     
                  </span>
                </span>
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  /&gt;
                </span>
              </div>
            </div>
            <button
              className="css-1naocv3 eva467m0"
              onClick={[Function]}
            >
              <div
                className="css-lvl6aa eva467m1"
              >
                <div
                  style={
                    Object {
                      "marginBottom": 6,
                    }
                  }
                >
                  Copied!
                </div>
                <div>
                  Copy
                </div>
              </div>
            </button>
          </pre>
        </div>
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Prop Types
          </h1>
          <div>
            <h2
              style={
                Object {
                  "margin": "20px 0 0 0",
                }
              }
            >
              "
              Loader
              " Component
            </h2>
            <small>
              No propTypes defined!
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Loader error 1`] = `
<div>
  <div
    style={
      Object {
        "position": "relative",
        "zIndex": 0,
      }
    }
  >
    <div
      className="custom-loader center"
    >
      <svg
        className="icon error"
        height="126"
        viewBox="0 0 126 126"
        width="126"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fillRule="evenodd"
        >
          <g
            className="arrow1"
          >
            <path
              d="M117.052 38.72C107.787 17.665 87.042 3 62.938 3c-32.74 0-59.28 27.052-59.28 60.422"
              strokeWidth="5.4"
            />
            <path
              d="M122.334 18.285l-3.32 20.692-18.506-5.075"
              strokeLinejoin="round"
              strokeWidth="6"
            />
          </g>
          <g
            className="arrow2"
          >
            <path
              d="M8.948 87.28C18.213 108.335 38.958 123 63.062 123c32.74 0 59.28-27.052 59.28-60.422"
              strokeWidth="5.4"
            />
            <path
              d="M3.666 107.715l3.32-20.692 18.506 5.075"
              strokeLinejoin="round"
              strokeWidth="6"
            />
          </g>
          <text
            className="mark"
          >
            <tspan
              x="46.239"
              y="84.175"
            >
              !
            </tspan>
          </text>
        </g>
      </svg>
      <p
        className="description center"
      >
        Il y a une erreur de chargement de l’analyse
      </p>
    </div>
  </div>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "#28c",
        "border": "none",
        "borderRadius": "0 0 0 5px",
        "color": "#fff",
        "cursor": "pointer",
        "display": "block",
        "fontFamily": "sans-serif",
        "fontSize": "12px",
        "padding": "5px 15px",
        "position": "fixed",
        "right": 0,
        "top": 0,
      }
    }
    type="button"
  >
    Show Info
  </button>
  <div
    style={
      Object {
        "background": "white",
        "bottom": 0,
        "display": "none",
        "left": 0,
        "overflow": "auto",
        "padding": "0 40px",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999,
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#28c",
          "border": "none",
          "borderRadius": "0 0 0 5px",
          "color": "#fff",
          "cursor": "pointer",
          "display": "block",
          "fontFamily": "sans-serif",
          "fontSize": "12px",
          "padding": "5px 15px",
          "position": "fixed",
          "right": 0,
          "top": 0,
        }
      }
      type="button"
    >
      ×
    </button>
    <div>
      <div
        style={
          Object {
            "WebkitFontSmoothing": "antialiased",
            "backgroundColor": "#fff",
            "border": "1px solid #eee",
            "borderRadius": "2px",
            "color": "#444",
            "fontFamily": "-apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", \\"Arial\\", sans-serif",
            "fontSize": "15px",
            "fontWeight": 300,
            "lineHeight": 1.45,
            "marginBottom": "20px",
            "marginTop": "20px",
            "padding": "20px 40px 40px",
          }
        }
      >
        <div
          style={
            Object {
              "borderBottom": "1px solid #eee",
              "marginBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <h1
            style={
              Object {
                "fontSize": "35px",
                "margin": 0,
                "padding": 0,
              }
            }
          >
            Loader
          </h1>
          <h2
            style={
              Object {
                "fontSize": "22px",
                "fontWeight": 400,
                "margin": "0 0 10px 0",
                "padding": 0,
              }
            }
          >
            error
          </h2>
        </div>
        
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Story Source
          </h1>
          <pre
            className="css-r8d96o eyvlbql0"
          >
            <div>
              <div
                style={
                  Object {
                    "paddingLeft": 18,
                    "paddingRight": 3,
                  }
                }
              >
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  &lt;
                  Loader
                </span>
                <span>
                  <span>
                     
                    <span
                      style={Object {}}
                    >
                      type
                    </span>
                    <span>
                      =
                      <span
                        style={Object {}}
                      >
                        "
                        <span
                          style={
                            Object {
                              "color": "#22a",
                              "wordBreak": "break-word",
                            }
                          }
                        >
                          ERROR
                        </span>
                        "
                      </span>
                    </span>
                     
                  </span>
                </span>
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  /&gt;
                </span>
              </div>
            </div>
            <button
              className="css-1naocv3 eva467m0"
              onClick={[Function]}
            >
              <div
                className="css-lvl6aa eva467m1"
              >
                <div
                  style={
                    Object {
                      "marginBottom": 6,
                    }
                  }
                >
                  Copied!
                </div>
                <div>
                  Copy
                </div>
              </div>
            </button>
          </pre>
        </div>
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Prop Types
          </h1>
          <div>
            <h2
              style={
                Object {
                  "margin": "20px 0 0 0",
                }
              }
            >
              "
              Loader
              " Component
            </h2>
            <small>
              No propTypes defined!
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Storyshots Loader playground 1`] = `
<div>
  <div
    style={
      Object {
        "position": "relative",
        "zIndex": 0,
      }
    }
  >
    <div
      className="custom-loader center"
    >
      <svg
        className="icon spinner"
        viewBox="0 0 66 66"
      >
        <circle
          className="path"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          strokeLinecap="round"
          strokeWidth="6"
        />
      </svg>
      <p
        className="description center"
      >
        Chargement en cours...
      </p>
    </div>
  </div>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "#28c",
        "border": "none",
        "borderRadius": "0 0 0 5px",
        "color": "#fff",
        "cursor": "pointer",
        "display": "block",
        "fontFamily": "sans-serif",
        "fontSize": "12px",
        "padding": "5px 15px",
        "position": "fixed",
        "right": 0,
        "top": 0,
      }
    }
    type="button"
  >
    Show Info
  </button>
  <div
    style={
      Object {
        "background": "white",
        "bottom": 0,
        "display": "none",
        "left": 0,
        "overflow": "auto",
        "padding": "0 40px",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999,
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#28c",
          "border": "none",
          "borderRadius": "0 0 0 5px",
          "color": "#fff",
          "cursor": "pointer",
          "display": "block",
          "fontFamily": "sans-serif",
          "fontSize": "12px",
          "padding": "5px 15px",
          "position": "fixed",
          "right": 0,
          "top": 0,
        }
      }
      type="button"
    >
      ×
    </button>
    <div>
      <div
        style={
          Object {
            "WebkitFontSmoothing": "antialiased",
            "backgroundColor": "#fff",
            "border": "1px solid #eee",
            "borderRadius": "2px",
            "color": "#444",
            "fontFamily": "-apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", \\"Arial\\", sans-serif",
            "fontSize": "15px",
            "fontWeight": 300,
            "lineHeight": 1.45,
            "marginBottom": "20px",
            "marginTop": "20px",
            "padding": "20px 40px 40px",
          }
        }
      >
        <div
          style={
            Object {
              "borderBottom": "1px solid #eee",
              "marginBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <h1
            style={
              Object {
                "fontSize": "35px",
                "margin": 0,
                "padding": 0,
              }
            }
          >
            Loader
          </h1>
          <h2
            style={
              Object {
                "fontSize": "22px",
                "fontWeight": 400,
                "margin": "0 0 10px 0",
                "padding": 0,
              }
            }
          >
            playground
          </h2>
        </div>
        
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Story Source
          </h1>
          <pre
            className="css-r8d96o eyvlbql0"
          >
            <div>
              <div
                style={
                  Object {
                    "paddingLeft": 18,
                    "paddingRight": 3,
                  }
                }
              >
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  &lt;
                  Loader
                </span>
                <span>
                  <span>
                     
                    <span
                      style={Object {}}
                    >
                      type
                    </span>
                    <span>
                      =
                      <span
                        style={Object {}}
                      >
                        "
                        <span
                          style={
                            Object {
                              "color": "#22a",
                              "wordBreak": "break-word",
                            }
                          }
                        >
                          LOADING
                        </span>
                        "
                      </span>
                    </span>
                     
                  </span>
                </span>
                <span
                  style={
                    Object {
                      "color": "#444",
                    }
                  }
                >
                  /&gt;
                </span>
              </div>
            </div>
            <button
              className="css-1naocv3 eva467m0"
              onClick={[Function]}
            >
              <div
                className="css-lvl6aa eva467m1"
              >
                <div
                  style={
                    Object {
                      "marginBottom": 6,
                    }
                  }
                >
                  Copied!
                </div>
                <div>
                  Copy
                </div>
              </div>
            </button>
          </pre>
        </div>
        <div>
          <h1
            style={
              Object {
                "borderBottom": "1px solid #EEE",
                "fontSize": "25px",
                "margin": "20px 0 0 0",
                "padding": "0 0 5px 0",
              }
            }
          >
            Prop Types
          </h1>
          <div>
            <h2
              style={
                Object {
                  "margin": "20px 0 0 0",
                }
              }
            >
              "
              Loader
              " Component
            </h2>
            <small>
              No propTypes defined!
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
